### PR TITLE
v4.0.x: VERSION: make safe for C++ compilers

### LINKED ISSUE #9122: Issue with Openmpi compilation on case-insensitive systems with clang
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

Tried with 4.0.5, 4.0.6, 4.1.0 and 4.1.1.

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

From a tarball, with command

```shell
 ./configure --enable-cxx-exceptions --with-hwloc=internal --with-libevent=internal  --prefix=/Volumes/Data/ci/opt/clang_release/Openmpi CC=clang CXX=clang++ FC=/usr/local/Cellar/gcc/11.1.0_1/bin/gfortran CFLAGS='-m64  -O3 -DNDEBUG ' CXXFLAGS='-m64  -stdlib=libc++ -O3 -DNDEBUG ' FCFLAGS='-m64   -O3 -DNDEBUG ' LDFLAGS='-m64  -O3 -DNDEBUG ' --enable-mpi-cxx
```

### Please describe the system on which you are running

Tried both on macOS Catalina and macOS BigSur; computers were few years old MacPro.

## Details of the problem

When compiling Openmpi, I get messages like:

```shell
In file included from comm.cc:26:
In file included from ../../../ompi/mpi/cxx/mpicxx.h:42:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/map:481:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__tree:14:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/iterator:419:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__functional_base:14:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/type_traits:417:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/cstddef:37:
../../../version:1:3: error: invalid preprocessing directive
# Copyright (c) 2008      Sun Microsystems, Inc.  All rights reserved.
  ^
../../../version:2:3: error: invalid preprocessing directive
# Copyright (c) 2008-2019 Cisco Systems, Inc.  All rights reserved.
  ^
../../../version:3:3: error: invalid preprocessing directive
# Copyright (c) 2011      NVIDIA Corporation.  All rights reserved.
  ^
../../../version:4:3: error: invalid preprocessing directive
# Copyright (c) 2013      Mellanox Technologies, Inc.
  ^
../../../version:5:27: error: invalid preprocessing directive
#                         All rights reserved.
                          ^
../../../version:6:3: error: invalid preprocessing directive
In file included from file.cc# Copyright (c) 2017      Los Alamos National Security, LLC.  All rights:
18:
  ^In file included from 
../../../ompi/mpi/cxx/mpicxx.h:42:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/map../../../version::4817:
:In file included from 27/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/v1/__tree:: 14:
```

I was a bit astonished, as compiling on my MacBookPro with BigSur went fine... After investigating a bit I think this is the same issue [reported here in another project](https://github.com/spack/spack/issues/11515) (I don't know if they mentioned it to you or not).

In particular, I think [this comment](https://github.com/spack/spack/issues/11515#issuecomment-495626063) nails it: the error message is really the first lines of the _VERSION_ file, and looking into _cstddef_ there is an `#include <version>` which is in all likelihood poorly interpreted... This would explain why I wasn't stuck on my personal computer: I chose a case sensitive filesystem, contrary to both MacPro on which I tried to compile Openmpi later.

The comment cited above suggests it is because _stage_ directory is put before the system includes, so the fix may not be difficult to provide (but I must admit I'm not proficient with autotools...).

### PR DESCRIPTION
**Pull request by jsquyres**
When compiling the MPI C++ bindings on a case-insensitive filesystem,
we #include various STL header files.  At least some versions of Clang
on MacOS have `#include <version>` in their STL header files, which
ends up including the Open MPI VERSION file.

The Open MPI VERSION file is executed as a Bourne shell script during
configure.  We could rename VERSION `-->` VERSION.sh, but this would
touch a lot of places, and doesn't seem worth it (remember: the MPI
C++ bindings have been removed from the Open MPI master and upcoming
v5.0.0 release).

This commit therefore does something small/simple to make the existing
VERSION file also safe to #include as a C/C++ header file.  It
basically uses "#if 0" to comment out the main body of the file, and
then `#include_next <version>` to get the "real" `<version>` file.

This commit is a cherry pick from the v4.1.x branch (since the issue
does not exist on master).

Thanks to @srpgilles for raising the issue.

Signed-off-by: Jeff Squyres <jsquyres@cisco.com>
(cherry picked from commit 4dd5a47a9554b85ca0d6b3d3f5814eaa87198274)

This is a cherry pick from the v4.1.x branch.

Fixes #9122

### CODE REVIEW

**Review by bwbarrett — APPROVED** [review]

**Comment by hjelmn** [comment]
WG21 didn't think this through at all :).

**Comment by ggouaillardet** [comment]
This was discussed here in a more colorful language before SC'19, and I can tell you first hand the C++ standard committee fully understood the implications of `#include <version>` and **chose** not to care and move forward anyway.

**Comment by jsquyres** [comment]
To be fair, we could also rename `VERSION` to be `VERSION.sh`; that would also fix the problem.  It's just a little bigger change than we're willing to make on a legacy release branch (because there's multiple places that read that file that would need to be updated).  Fortunately, this hack is available that also solves the problem via a smaller change.
